File: homing-workspace/src/main/java/hue/captains/singapura/js/homing/workspace/events/WorkspaceEventLog.java
package hue.captains.singapura.js.homing.workspace.events;

import java.io.IOException;
import java.io.InputStream;
import java.io.UncheckedIOException;
import java.nio.charset.StandardCharsets;

/**
 * Drop-in bundler that returns the complete JS source for the RFC 0030
 * workspace event log as a single concatenated string, ready for inclusion
 * in a workspace chrome's {@code bodyJs()}.
 *
 * <p>Phase 1 — record-only. The chrome calls {@code WorkspaceEventLog.attach(...)}
 * once per workspace instance with a {@code (workspaceKind, workspaceId)} pair,
 * then {@code emit(name, payload)} on every state-affecting action. Events
 * append to IndexedDB under a per-instance composite key; recovery still
 * uses RFC 0029 snapshots through Phase 2. Replay-from-events lands in
 * Phase 2 once we've verified the event vocabulary covers what the chrome
 * actually does.</p>
 *
 * <p>The bundle contains, in dependency order:</p>
 * <ol>
 *   <li>{@code EventLogStore.js} — IndexedDB wrapper (append + range query
 *       + clear), keyed by {@code (kind, workspaceId, seq)}.</li>
 *   <li>{@code WorkspaceEventLog.js} — high-level facade that binds the
 *       {@code (kind, workspaceId)} pair once and exposes {@code emit / query /
 *       clear}.</li>
 * </ol>
 *
 * <h2>Usage in a workspace chrome's {@code bodyJs()}</h2>
 *
 * <pre>{@code
 * return List.of(
 *     WorkspaceEventLog.allJs(),       // prepend the bundle
 *     "// rest of chrome body...",
 *     "var eventLog = WorkspaceEventLog.attach({",
 *     "    workspaceKind: 'AnimalsPlayground',",
 *     "    workspaceId:   computePlaceholderInstanceUuid('AnimalsPlayground')",
 *     "});",
 *     "eventLog.emit('TabAdded', { paneId: 'tl', tabId: '...', widgetKind: '...' });",
 *     ...
 * );
 * }</pre>
 *
 * <p>Per-workspace scoping aligns with RFC 0031 from day one — events are
 * already segregated by {@code (kind, workspaceId)} so the multi-workspace
 * promotion is a no-op at the storage layer. Single-workspace use today
 * passes a deterministic placeholder UUID derived from the kind name.</p>
 *
 * @since RFC 0030 Phase 1 — record-only event log.
 */
public final class WorkspaceEventLog {

    private WorkspaceEventLog() {}

    /**
     * The complete event-log JS bundle as a single string. Cached on first
     * call — the bundle is pure resource concatenation and deterministic.
     */
    public static String allJs() {
        return BUNDLE;
    }

    private static final String BUNDLE = build();

    private static String build() {
        var sb = new StringBuilder();
        sb.append("// ─── RFC 0030 event log bundle ────────────────────────────────────\n");
        appendResource(sb, "EventLogStore.js");
        appendResource(sb, "WorkspaceEventLog.js");
        return sb.toString();
    }

    private static void appendResource(StringBuilder sb, String fileName) {
        // Resources live alongside this Java class at
        // /homing/js/hue/captains/singapura/js/homing/workspace/events/
        String path = "/homing/js/hue/captains/singapura/js/homing/workspace/events/" + fileName;
        try (InputStream in = WorkspaceEventLog.class.getResourceAsStream(path)) {
            if (in == null) {
                throw new IllegalStateException(
                        "WorkspaceEventLog: missing classpath resource " + path);
            }
            sb.append(new String(in.readAllBytes(), StandardCharsets.UTF_8));
            sb.append('\n');
        } catch (IOException e) {
            throw new UncheckedIOException("Failed to read " + path, e);
        }
    }
}

File: homing-workspace/src/main/resources/homing/js/hue/captains/singapura/js/homing/workspace/events/WorkspaceEventLog.js
// =============================================================================
// WorkspaceEventLog — per-workspace facade over EventLogStore.
//
// RFC 0035 P2 — refactored from `Object.freeze({attach: ...})` IIFE to class
// form. The class IS the substrate-free EventLog<WorkspaceEventPayload>
// contract from the Java contract package — bound to one (kind, workspaceId)
// pair at construction time, exposes append / query / clear as the public
// method set.
//
// Per-workspace scoping per RFC 0031 multi-workspace, the (kind, workspaceId)
// pair keys every IDB row. Single-workspace use today passes a deterministic
// placeholder UUID (see WorkspaceInstanceId.placeholderFor on the Java side).
//
// Usage in a chrome's bodyJs():
//
//   const eventLog = new WorkspaceEventLog({
//       workspaceKind: "AnimalsPlayground",
//       workspaceId:   "<deterministic-placeholder-uuid>"
//   });
//   eventLog.append("TabAdded", { paneId: "tl", tabId: "...", widgetKind: "..." });
//
// Phase 1 record-only — errors during recording land in console.warn and
// the live UI never blocks on storage.
// =============================================================================

class WorkspaceEventLog {

    constructor(opts) {
        if (!opts || typeof opts.workspaceKind !== "string" || opts.workspaceKind.length === 0) {
            throw new TypeError("WorkspaceEventLog: opts.workspaceKind required (non-empty string)");
        }
        if (typeof opts.workspaceId !== "string" || opts.workspaceId.length === 0) {
            throw new TypeError("WorkspaceEventLog: opts.workspaceId required (non-empty string)");
        }
        this.workspaceKind = opts.workspaceKind;
        this.workspaceId   = opts.workspaceId;
        this._store        = opts.store || createEventLogStore();

        // Monotonic local counter for the session — useful for ordering
        // events emitted in the same millisecond. The authoritative seq
        // comes back from IndexedDB; this is just for in-session tracing.
        this._localCount = 0;
    }

    /**
     * Append one event. Returns a Promise that resolves to the assigned
     * seq number. Fire-and-forget on the recording path; the promise is
     * there for tests that want the seq + chrome callsites that need it
     * for cadence tracking (RFC 0034).
     *
     * Conforms to Java EventLog<P>.append(EventName, P): the JS arguments
     * mirror the contract's parameter order. EventName arrives as a string
     * (the typed payload's NAME constant on the Java side; the JS author
     * holds the same convention).
     */
    append(name, payload) {
        this._localCount++;
        return this._store.append(this.workspaceKind, this.workspaceId, name, payload || {})
                   .catch(e => {
                       console.warn("[WorkspaceEventLog] append failed for", name, "—",
                                    e && e.message ? e.message : e);
                       throw e;
                   });
    }

    /**
     * Range-query this workspace's events. Optional { fromSeq, limit }.
     * Conforms to Java EventLog<P>.query(EventQuery) with the EventQuery
     * record's fields passed as a plain object.
     */
    query(opts) {
        return this._store.query(this.workspaceKind, this.workspaceId, opts);
    }

    /**
     * Drop every event for this workspace. Used by Reset State.
     * Resolves to the count deleted.
     */
    clear() {
        return this._store.clear(this.workspaceKind, this.workspaceId);
    }

    /** Diagnostics: how many events have been append()ed this session. */
    localEmitCount() { return this._localCount; }

    // ─── Legacy emit() alias ─────────────────────────────────────────────
    // Pre-RFC-0035 chromes called eventLog.emit(name, payload). The method
    // is identical to append(); kept until all chromes migrate to the
    // contract-aligned name.

    /** @deprecated use {@link append}. */
    emit(name, payload) { return this.append(name, payload); }

    /**
     * Legacy {@code WorkspaceEventLog.attach({...})} call shape preserved
     * as a static factory. Pre-RFC-0035 chromes used the IIFE form; this
     * static lets them keep working unchanged. New code uses
     * {@code new WorkspaceEventLog(opts)} directly.
     *
     * @deprecated use {@code new WorkspaceEventLog(opts)}.
     */
    static attach(opts) { return new WorkspaceEventLog(opts); }
}
